controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -946,118 +946,169 @@ private boolean isSubscriberToNonSubscriberTransition(final ReaderGroupConfigRec
-                        } else {
-                            // if retained size is greater than max allowed, then truncate the stream at streamcut
-                            // from retention set that matches the retention policy size bound.
-                            return maxBound.get().map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor)
-                            .thenApply(StreamCutRecord::getStreamCut)
-                            .thenCompose(maxRecord -> {
-                            // if max record is strictly greater than lowerbound then we can truncate at max record
-                            return streamMetadataStore.streamCutStrictlyGreaterThan(
-                            scope, stream, maxRecord, lowerBound, context, executor)
-                            .thenApply(gt -> {
-                            if (gt) {
-                                return maxRecord;
-                              } else {
-                                return lowerBound;
-                              }
-                            });
-                            })).orElse(CompletableFuture.completedFuture(null));
+                        } else { // greater than max. truncate at max. 
+                            // let there be data loss. its a lagging reader.. but if there is no streamcut in 
+                            // retentionset that satisfied the min and max criteria, then we should simply truncate at 
+                            // least at the lowerbound 
+                            return Optional.ofNullable(limits.getKey())
+                                           .filter(x -> currentSize - x.getRecordingSize() < retainedSizeLB)
+                                           .map(x -> streamMetadataStore.getStreamCutRecord(scope, stream, x, context, executor)
+                                                                        .thenApply(StreamCutRecord::getStreamCut))
+                                    .orElse(CompletableFuture.completedFuture(lowerBound));
                         }
-                      }
+                    }
                 });
     }
 
     private CompletableFuture<Map<Long, Long>> getTruncationStreamCutByTimeLimit(String scope, String stream, OperationContext context,
                                                                                  RetentionPolicy policy, RetentionSet retentionSet,
                                                                                  Map<Long, Long> lowerBound) {
-        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits =
-                getBoundStreamCuts(policy, retentionSet);
-        // if subscriber lowerbound is ahead of streamcut corresponding to the max time and is behind stream cut for min time
+        long currentTime = retentionClock.get().get();
+
+        // we get the streamcuts from retentionset that satisfy the min and max bounds with min pointing to most recent 
```

Comment:
This line essentially is the same as line #955 in getRetentionStreamCutsBySizeLimit().
But the  comments don't seem to match....it may be good to just explain at the method definition of getBoundStreamCuts() what this method does and what values it returns.

Reply:
the comment is same as  line 955 because they are both calling the same method with different delta functions for size and time based retained value computation respectively.
hence identical comments.  
As suggested i will add more comments inside the getBoundStreamCuts method to make it clear what it is computing.